core/local/local.go

```diff
@@ -23,527 +23,363 @@ package local
+// out channel.
+func (e *Executor) Run(ctx context.Context, engineOut chan<- stats.SampleContainer) error {
+	schedulersCount := len(e.schedulers)
+	logger := e.logger.WithField("phase", "local-executor-run")
+	e.initProgress.Modify(pb.WithConstLeft("Run"))
 
-func (e *Executor) SetStages(s []lib.Stage) {
-	e.stages = s
-}
-
-func (e *Executor) GetIterations() int64 {
-	return atomic.LoadInt64(&e.iters)
-}
-
-func (e *Executor) GetEndIterations() null.Int {
-	v := atomic.LoadInt64(&e.endIters)
-	if v < 0 {
-		return null.Int{}
-	}
-	return null.IntFrom(v)
-}
-
-func (e *Executor) SetEndIterations(i null.Int) {
-	if !i.Valid {
-		i.Int64 = -1
+	if e.state.IsPaused() {
+		logger.Debug("Execution is paused, waiting for resume or interrupt...")
+		e.initProgress.Modify(pb.WithConstProgress(1, "paused"))
+		select {
+		case <-e.state.ResumeNotify():
+			// continue
+		case <-ctx.Done():
+			return nil
+		}
 	}
-	e.Logger.WithField("i", i.Int64).Debug("Local: Setting end iterations")
-	atomic.StoreInt64(&e.endIters, i.Int64)
-}
 
-func (e *Executor) GetTime() time.Duration {
-	return time.Duration(atomic.LoadInt64(&e.time))
-}
-
-func (e *Executor) GetEndTime() types.NullDuration {
-	v := atomic.LoadInt64(&e.endTime)
-	if v < 0 {
-		return types.NullDuration{}
-	}
-	return types.NullDurationFrom(time.Duration(v))
-}
+	e.state.MarkStarted()
+	defer e.state.MarkEnded()
+	e.initProgress.Modify(pb.WithConstProgress(1, "running"))
 
-func (e *Executor) SetEndTime(t types.NullDuration) {
-	if !t.Valid {
-		t.Duration = -1
-	}
-	e.Logger.WithField("d", t.Duration).Debug("Local: Setting end time")
-	atomic.StoreInt64(&e.endTime, int64(t.Duration))
-}
+	logger.WithFields(logrus.Fields{"schedulersCount": schedulersCount}).Debugf("Start of test run")
 
-func (e *Executor) IsPaused() bool {
-	e.pauseLock.RLock()
-	defer e.pauseLock.RUnlock()
-	return e.pause != nil
-}
+	runResults := make(chan error, schedulersCount) // nil values are successful runs
 
-func (e *Executor) SetPaused(paused bool) {
-	e.Logger.WithField("paused", paused).Debug("Local: Setting paused")
-	e.pauseLock.Lock()
-	defer e.pauseLock.Unlock()
+	runCtx, cancel := context.WithCancel(ctx)
+	defer cancel() // just in case, and to shut up go vet...
 
-	if paused && e.pause == nil {
-		e.pause = make(chan interface{})
-	} else if !paused && e.pause != nil {
-		close(e.pause)
-		e.pause = nil
+	// Run setup() before any schedulers, if it's not disabled
+	if !e.options.NoSetup.Bool {
+		logger.Debug("Running setup()")
+		e.initProgress.Modify(pb.WithConstProgress(1, "setup()"))
+		if err := e.runner.Setup(runCtx, engineOut); err != nil {
+			logger.WithField("error", err).Debug("setup() aborted by error")
+			return err
+		}
 	}
-}
-
-func (e *Executor) GetVUs() int64 {
-	return atomic.LoadInt64(&e.numVUs)
-}
+	e.initProgress.Modify(pb.WithHijack(e.getRunStats))
+
+	runCtxDone := runCtx.Done()
+	runScheduler := func(sched lib.Scheduler) {
```

Comment:
Maybe this can be moved in a method?
The freevars reported are :
```
  1 core/local/local.go|295 col 1| Free identifiers:
  2 core/local/local.go|255 col 45| var engineOut chan<- github.com/loadimpact/k6/stats.SampleContainer
  3 core/local/local.go|257 col 2| var logger.WithFields func(fields github.com/loadimpact/k6/vendor/github.com/sirupsen/logrus.Fields) *github.com/loadimpact/k6/vendor/github.com/sirupsen/logrus.Entry
  4 core/local/local.go|279 col 2| var runCtx context.Context
  5 core/local/local.go|293 col 2| var runCtxDone <-chan struct{}
  6 core/local/local.go|277 col 2| var runResults chan error
  7 core/local/local.go|294 col 23| var sched.GetConfig func() github.com/loadimpact/k6/lib.SchedulerConfig
  8 core/local/local.go|294 col 23| var sched.GetProgress func() *github.com/loadimpact/k6/ui/pb.ProgressBar
  9 core/local/local.go|294 col 23| var sched.Run func(ctx context.Context, engineOut chan<- github.com/loadimpact/k6/stats.SampleContainer) error
```
So `engineOut`, `runCtx`(Done can be taken from it), `runResults` and `sched`.

Reply:
You missed `logger`, so 5 arguments in total... would taking this out as a method really make the code more readable? I don't mind doing it, I'm just unsure if having to jump between the two would be easier when reading the code.

Reply:
given that you move 30+ lines of code out of a 100 line function ... yes :)

Reply:
Resolved by https://github.com/loadimpact/k6/pull/1007/commits/a3da2afd1ff4d8592f12040750e69498f5fc015a